### ISSUE TITLE
Disabling update check on regular requests?

### ISSUE THREAD

**Issue by khromov** [afragen/github-updater#242]
When a site has not been visited for a while, GitHub Updater decides to load external data from the GitHub API in the same process as the page load. Dozens of HTTP requests are sent out and the user is just waiting.

Here is what that looks like.
![github-updater-damn](https://cloud.githubusercontent.com/assets/1207507/7948463/6e3d78da-0984-11e5-878d-60edac20264f.png)

25 HTTP requests, 22.67 seconds TTFB (!)

The issues happens when creating the plugin instance and calling get_remote_info():
https://github.com/afragen/github-updater/blob/develop/src/GitHub_Updater/Plugin.php#L64

It would be much better to register a proper WP Cron job that would handle this. In fact, we can see that a WP Cron job also starts when a site hasn't been visited for a while:
![github-updater-damn2](https://cloud.githubusercontent.com/assets/1207507/7948547/11f2c156-0985-11e5-9669-5ae32de99932.png)

This issue is quite severe if you have a lot of GitHub Updater-enabled plugins, I can't currently use it because forcing a random visitors to wait 15-30 seconds for a page load isn't acceptable. Hoping it can be resolved.

**Comment by afragen** [comment]
This only happens when an admin user visits the site and all data is cached for 12 hours. Regular users won't see a performance hit. Test with any of the site speed tests.

I made quick stab at using cron and while I got a cron job set up it wouldn't show updates. I know there must be something more to this but I haven't had the time to figure it out.

**Bottom line:** no regular users will be impacted by this plugin. Please refer to the following code.
https://github.com/afragen/github-updater/blob/develop/src/GitHub_Updater/Base.php#L105-L115

Admin users will only see this once every 12 hours, unless they use the 'Check Again' button.

**Comment by khromov** [comment]
Thanks for the answer. I'll give the WP Cron method a stab and submit a PR, because that's the way WP checks for updates internally, so maybe we can even piggyback on that CRON call.

**Comment by afragen** [comment]
That would be great. If love to see the performance differences. As long as it's functionally identical we could move towards we-cron.

**Comment by yafred** [comment]
@khromov Hi, what is the plugin you're using to analyze performances ?

**Comment by khromov** [comment]
@yafred The plugin in the screenshot is Query Monitor:  
https://wordpress.org/plugins/query-monitor/

@afragen I started doing the Cron refactoring. Not quite ready yet but expect a PR soon. :-)

**Comment by afragen** [comment]
@khromov excellent! Please try to ensure that using the 'Check Again' button will both clear the transients and reset the cron job to run at next page load.

I've been refactoring the renaming code but somehow what works locally doesn't work in the server for multiple selections. :tired_face: 

BTW, @yafred Query Monitor is a **must use** for development.

**Comment by afragen** [comment]
@khromov you might need to update your fork against `develop`. I had to revert the renaming refactor as above.

**Comment by unetics** [comment]
+1 Would love to see this as i am finding the same problems and think cron would be a great solution.

**Comment by afragen** [comment]
@unetics please remember that this "hit" only happens once every 12 hours and only for admin-level users.

**Comment by afragen** [comment]
@khromov any progress on the PR for using wp_cron?

**Comment by khromov** [comment]
@afragen Started to do the changes but didn't finish it. Not far off though, I'll submit a PR in the coming days!

**Comment by afragen** [comment]
Great! Make sure you're updated to latest `develop` branch too. It's been changing since you've started. :wink:

**Comment by khromov** [comment]
@afragen Left you something in #270 :smile_cat:

**Comment by afragen** [comment]
Thanks very much! :100:

**Comment by afragen** [comment]
@khromov, it might be easier to move the conversation regarding this here instead of over at the closed PR. I just think it might be easier to find.

@unetics if you want to test the latest `develop` branch you'll find what I think is going to be the end result of this. Only on certain admin pages in the dashboard is the full plugin API called.

**Comment by afragen** [comment]
@khromov how's the vacation? Have you had an opportunity to test or were you kidding about the time to code on vacation? If so, I missed it as that's when I do have more opportunity to code.

**Comment by khromov** [comment]
@afragen No kidding, just doing other stuff, we have a big [WordCamp in Sweden](https://norrkoping.wordcamp.org/2015/) coming up soon, consider yourself warmly invited. :smile: 

I have looked at the code, it does appear to function properly. One thing caught my eye:

https://github.com/afragen/github-updater/blob/develop/src/GitHub_Updater/Base.php#L152

Why do we need to enable `$force_meta_update` on AJAX requests? Consider an admin user entering his dashboard and some AJAX request spinning off to fetch data in a third party plugin/theme. That will trigger remote data fetching and the AJAX request will be blocked until GitHub responds to all requests. If this is to amend some third party integration I'm sure we can find a better way. AJAX content is still content the user expects on a page load so it wouldn't be great forcing them to wait for it.

Other than that, looks good and we should try to ship it. Improvements can come later. :)

In the long run though, it could probably be improved from an architectural standpoint. 

I'd prefer to see `Plugin::$object` and `Theme::$object` always be initialized (and not just on `$admin_pages` as today), even if they are "empty" (ie. the update data has gone stale or has not been loaded yet). 

I made a proposition for a refactoring the update system that you can find here, together with description on which changes are made and why:
https://github.com/afragen/github-updater/pull/279/files?w=1

Would love to hear your feedback.

**Comment by afragen** [comment]
@khromov I hope #wcnkpg was a blast!

I've done a bunch of refactoring in the `develop` branch. Mostly to bring all the plugin code into Plugin.php and all the theme code into Theme.php.

Filters are called in their own method and I've fixed a couple of bugs/issues.

I've got another branch working to bypass `wp_remote_get` calls from other plugins, currently only Modern Tribe's plugins, that run a `wp_remote_get` call every time either `wp_update_plugins()` or `pre_set_site_transient_update_plugins` is run. Can't quite figure out which, but I bypass those HTTP requests for 12 hours.

Let me know what you think.
